torch/csrc/jit/script/compiler.cpp

```diff
@@ -34,23 +34,127 @@ using ListAttributeMap = std::unordered_map<
     std::string,
     std::pair<const std::vector<double>, std::string>>;
 
+// Keep track of environment as we descend down nested control
+// structures.
+struct Environment {
+  Environment(Block* b, std::shared_ptr<Environment> next = nullptr)
+      : b(b), next(next) {}
+
+  std::vector<std::string> captured_inputs;
+  ValueTable value_table;
+  Block* b;
+
+  std::shared_ptr<Environment> next;
+
+  Value* findInThisFrame(const std::string& name) {
+    if (value_table.count(name)) {
+      return value_table.at(name);
+    }
+    return nullptr;
+  }
+
+  Value* findInParentFrame(const std::string& name) {
+    for (auto runner = next; runner; runner = runner->next) {
+      if (runner->value_table.count(name)) {
+        return runner->value_table.at(name);
+      }
+    }
+    return nullptr;
+  }
+
+  Value* createCapturedInput(const std::string& name) {
+    // Create the input
+    Value* new_input = b->addInput();
+
+    // Associate this name with this value
+    value_table[name] = new_input;
+
+    // List as a positional input
+    captured_inputs.push_back(name);
+
+    return new_input;
+  }
+
+  Symbol getBlockOwningKind() {
+    Symbol owning_kind = Symbol();
+    if (b->owningNode()) {
+      owning_kind = b->owningNode()->kind();
+    }
+    return owning_kind;
+  }
+
+  void setVar(const std::string& name, Value* value) {
+    if (!findInThisFrame(name) && findInParentFrame(name) &&
+        getBlockOwningKind() == kLoop)
+      createCapturedInput(name);
+    value_table[name] = value;
+  }
+
+  Value* getVar(const Ident& ident) {
+    return getVar(ident.name(), ident);
+  }
+
+  Value* getVar(const std::string& ident, const TreeView& tv) {
+    Value* retval = findInThisFrame(ident);
+
+    if (!retval && (retval = findInParentFrame(ident)) &&
+        getBlockOwningKind() == kLoop)
+      retval = createCapturedInput(ident);
```

Comment:
It is kind of weird that `getVar` mutates the state of the `Environment`, and further emphasizes the fact that this is not a generic `Environment` data structure, but simply a bookkeeping structure to help you do some of the transformations you need. What is the semantic meaning of `captured_inputs`? It is something to do with captured variables, but the relationship between the sequence of method calls to `Environment` and the value of this variable is murky.